daml/ContingentClaims/Lifecycle.daml

```diff
@@ -3,83 +3,120 @@
 -- SPDX-License-Identifier: Apache-2.0
 --
 
-{-# LANGUAGE MultiWayIf #-}
 {-# OPTIONS -Wno-deprecations #-} -- To supress 'Monad' warnings
 
 module ContingentClaims.Lifecycle (
-    lifecycle
-  , lifecycle'
+    settle
+  , settle'
   , Result(..)
 ) where
 
 import ContingentClaims.Claim (Claim, Claim(..), ClaimF(..))
 import ContingentClaims.Observation (Observation, eval)
 import ContingentClaims.Util (apoCataM, pruneZeros')
+import Daml.Control.Recursion (project, embed)
 import Daml.Control.Monad.Writer (WriterT, runWriterT)
 import Daml.Control.Monad.State (StateT(..), evalStateT)
 import Daml.Control.Monad.MonadWriter (MonadWriter(..))
 import Daml.Control.Monad.MonadState (MonadState(..), modify)
 import Daml.Control.Monad.Trans
 import Daml.Control.Arrow (Kleisli(..))
+import DA.Action ((>=>))
+import DA.Traversable (sequence)
 import Prelude hiding (sequence, mapA)
 
 type C a = Claim Observation Date a
 type F a = ClaimF Observation Date a
 
 deriving instance Eq a => Eq (C a)
 
--- | Returned from a lifecycle operation.
+-- | Returned from a `settle` operation.
 data Result a = Result with
   remaining : C a
     -- ^ the tree after lifecycled branches have been pruned.
   pending : [(Decimal, a)]
     -- ^ quantity/asset pairs requiring settlement.
 
--- | Lifecycle a Claim. Parameters:
--- 1. Function taking key and date, used to lookup an observable on the ledger. `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
--- 2. Function used when branching in an `Or`. Must return one of the arguments.
--- 3. A set of `Claim`s.
--- 4. A function which, given a date, lifecycles the third argument.
-lifecycle : (Eq a, CanAbort m)
+-- | Collect claims falling due into a list, and return the tree with those nodes pruned.
+-- `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
+-- The first argument is used to lookup the value of any `Observables`.
+-- Results in a function taking today's date, and returning the pruned tree + pending settlements.
+settle : (Eq a, CanAbort m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
   -> C a
   -> Date -> m (Result a)
-lifecycle spot choose
+settle spot
   = runKleisli
   . fmap (uncurry Result)
   . flip evalStateT 1.0
   . runWriterT
-  . apoCataM (pruneZeros') (lifecycle' spot choose)
+  . apoCataM pruneZeros' settleLazily'
+    where
+      settleLazily' = (lift . lift <$> lazily' spot)
+                      >=> (fmap distM (fmap (settle' spot . embed)) . sequence)
```

Comment:
This is where all the complexity is shoved away to. All this does is to compose `lazily'` and `settle'` and sequence the effects to have the right type signature: `C a -> M (F a (C a))`.

Reply:
Consider writing a combinator that does same as `apoCataM` but you can pass a `C a -> M (F a (C a))` as the second argument, and `lazily` is already pre-applied.

Reply:
Not quite what I wrote above, but I did implement a general `ghyloM`.